--- hardware/msm7k/liblights/Android.mk~	2011-07-08 23:02:40.000000000 +0200
+++ hardware/msm7k/liblights/Android.mk	2011-07-10 10:59:13.000000000 +0200
@@ -22,6 +22,7 @@
 
 LOCAL_PRELINK_MODULE := false
 LOCAL_MODULE_PATH := $(TARGET_OUT_SHARED_LIBRARIES)/hw
+LOCAL_MODULE_TAGS := optional
 
 LOCAL_SHARED_LIBRARIES := liblog
 
--- hardware/msm7k/libcopybit/Android.mk~	2011-07-08 23:02:40.000000000 +0200
+++ hardware/msm7k/libcopybit/Android.mk	2011-07-10 12:33:43.000000000 +0200
@@ -21,6 +21,7 @@
 include $(CLEAR_VARS)
 LOCAL_PRELINK_MODULE := false
 LOCAL_MODULE_PATH := $(TARGET_OUT_SHARED_LIBRARIES)/hw
+LOCAL_MODULE_TAGS := optional
 LOCAL_SHARED_LIBRARIES := liblog
 LOCAL_SRC_FILES := copybit.cpp
 LOCAL_MODULE := copybit.msm7k
--- hardware/msm7k/libgralloc/Android.mk~	2011-07-08 23:02:40.000000000 +0200
+++ hardware/msm7k/libgralloc/Android.mk	2011-07-10 10:59:28.000000000 +0200
@@ -20,6 +20,7 @@
 include $(CLEAR_VARS)
 LOCAL_PRELINK_MODULE := false
 LOCAL_MODULE_PATH := $(TARGET_OUT_SHARED_LIBRARIES)/hw
+LOCAL_MODULE_TAGS := optional
 LOCAL_SHARED_LIBRARIES := liblog libcutils
 
 LOCAL_SRC_FILES := 	\
