--- vendor/htc/buzz/buzz-vendor-blobs.mk~	2011-11-11 16:07:15.000000000 +0100
+++ vendor/htc/buzz/buzz-vendor-blobs.mk	2011-11-11 16:07:17.000000000 +0100
@@ -29,6 +29,7 @@
     vendor/htc/buzz/proprietary/libgps.so:system/lib/libgps.so \
     vendor/htc/buzz/proprietary/liboemcamera.so:system/lib/liboemcamera.so \
     vendor/htc/buzz/proprietary/libcamera.so:system/lib/libcamera.so \
+    vendor/htc/buzz/proprietary/libOmxCore.so:system/lib/libOmxCore.so \
     vendor/htc/buzz/proprietary/libOmxH264Dec.so:system/lib/libOmxH264Dec.so \
     vendor/htc/buzz/proprietary/libOmxMpeg4Dec.so:system/lib/libOmxMpeg4Dec.so \
     vendor/htc/buzz/proprietary/libOmxVidEnc.so:system/lib/libOmxVidEnc.so \
